### ISSUE TITLE
Add WASM CustomEvent Example

### ISSUE THREAD

**Issue by stevenhuyn** [rust-windowing/winit#2816]
The current CustomEvent example in the `./examples` folder doesn't show how to do it via WASM.

https://github.com/rust-windowing/winit/blob/master/examples/custom_events.rs

Could we add an example like this + a basic HTML file to show off how to fire custom events from the JS side into the WASM winit event loop?

https://github.com/rust-windowing/winit/discussions/2061#discussioncomment-5813125

**Comment by daxpedda** [comment]
I'm happy to review a PR adding this.

**Comment by stevenhuyn** [comment]
awesome I'll get on it

**Comment by daxpedda** [comment]
I honestly don't know.

One alternative could be to just use `web-sys` to insert the necessary JS and HTML you want. You could make a separate `script.js` file and use `include_str!()` to make it sensible. The obvious solution is to require some sort of Makefile that does things for you, but I don't think this is a good fit for the Winit project.

On a general note I believe this is more of a `wasm-bindgen` problem then a Winit one, so maybe enhancing the documentation there or adding an easy example (not necessary a Winit one) could be a better fit.

> we need to add a `wasm_bindgen` dependency

`wasm-bindgen` is already a dependency of Winit:
https://github.com/rust-windowing/winit/blob/3c3be71a775f772b50c2eca3de18853bb4fb9692/Cargo.toml#L152-L153

---

Let me know your plans before you proceed though. Don't want you to invest significant effort just to be declined later.

**Comment by maroider** [comment]
We could do a PR against [`cargo-run-wasm`](https://github.com/rukai/cargo-run-wasm) adding whatever we need to inject custom HTML and/or JS for a particular example. Or you could bait @rukai into doing it :P

**Comment by daxpedda** [comment]
:laughing:
But yeah, support in `cargo-run-wasm` would address the problem.

**Comment by stevenhuyn** [comment]
I'm seeing technically yeah we could just js inject it into the `cargo-run-wasm` call, probably with a js `include_str!` to keep it sane like you said.

https://github.com/rukai/cargo-run-wasm/blob/c380ca868f7764ff6753ba37696b62ff54d58496/src/lib.rs#LL164C3-L164C3
```rust
pub fn run_wasm_with_css(css: &str) {
    // validate css
    //
    // Someone could easily get around this with some extra spaces
    // but im not about to import regex or do a complicated implementation by hand.
    if css.contains("</style>") {
        panic!(
            "`</style>` detected in the css. This is disallowed to prevent injecting elements into the DOM."
        )
    }
```

**Comment by daxpedda** [comment]
Haha, no I meant with `web-sys`. Let's not circumvent `cargo-run-wasm`.
But building the feature into `cargo-run-wasm` sounds like the most sensible approach.

**Comment by stevenhuyn** [comment]
oh I was trying to bait the author

**Comment by rukai** [comment]
I've designed cargo-run-wasm such that the correct way to do custom html is to insert elements into the DOM via web-sys.
The reason I exposed a way to set css in cargo-run-wasm is so that I could make the documented default usage remove the webpage padding. (I didnt just apply that css implicitly since I wanted it to be explicit that we werent using default css)

Is using web-sys to insert html particularly troublesome here?

**Comment by daxpedda** [comment]
> Is using web-sys to insert html particularly troublesome here?

No! But the example proposed here is supposed to show JS users how to connect to Winit from their perspective. Having a clear `index.html` shows them how to do that. Doing it over `web-sys` moves it back to being Rust-centric.

It's not really a big deal, but would be an improvement nonetheless.

**Comment by rukai** [comment]
One possible approach you could take is pointing users to inspect the html file generated by cargo-run-wasm at `winit/target/wasm-examples/web/index.html`.
If you insert a `<script></script>` containing regular javascript via web-sys then the user gets a clear `index.html` demonstrating how regular javascript can interact.

**Comment by daxpedda** [comment]
Would be nice to actually commit the file too, then users can inspect it on GitHub as well. I think that could work for us, @stevenhuyn.

It's not 100% ideal I would argue, personally I prefer to avoid committing generated files and inserting `<script></script>` blocks isn't the smoothest experience either. But I completely understand the hesitation to add this feature to `cargo-run-wasm`. Thanks for the back-and-forth, it was very helpful!

**Comment by stevenhuyn** [comment]
thoughts on this thing I worked up? do you think the JS snippet is good enough explanation? 

Or should I web sys more the button creation, so the html can be more clear and thus can be added to example folder.

**Comment by stevenhuyn** [comment]
Made an attempt here: https://github.com/rukai/cargo-run-wasm/pull/36/files
Haven't updated the documentation yet.

A question I need to investigate is how to match a custom html file. My guess would be inside `run-wasm/src/main.rs` I would check the argument and match on the example name and get the html file or not and include it if matching.
